### PR TITLE
feat: adds idea board app with some challanges

### PR DESCRIPTION
**Pull request by gwidonw**
# ContentCal Frontend Challenge

## Our Stack

We are currently working with...

- [Angular 10+](https://angular.io/)
- [Typescript](https://www.typescriptlang.org/)
- [NX](https://nx.dev/) Workspace
- [NGXS](https://www.ngxs.io/) State Manager
- [Jest](https://jestjs.io/) for Unit Testing
- [Cypress](https://www.cypress.io/) for E2E testing
- [Storybook](https://storybook.js.org/) as a rapid development tool and component library
- SCSS - We currently support the latest 2 versions of Edge, Chrome, Firefox and Safari

## What we're looking for

- A stylish solution with unit test coverage
- Clean, concise code
- an appreciation of design

## Run the project

1. Run `npm install`
2. Run `npm start`
3. Enter `http://localhost:4200`

### Running unit tests

- Run `npm run affected:test` to execute the unit tests via Jest on affected files only.
- Run `npm run affected:test --all` to execute the unit tests via Jest on entire project.

### Running linters

- Run `npm run affected:lint` to execute linter on affected files only.
- Run `npm run affected:lint --all` to execute linter on entire project.

## Challange

The Front-end challenge is to improve an idea board where you can create, view, update and delete ideas.
There is a starter app built using NX or you can choose you use your own boilerplate.

idea-board

### Challange 1:

Oops, seems like you can access our idea list even if you are not logged in just by entering `http://localhost:4200/ideas`. This module should be accessed only by logged-in users. Could you please protect it somehow?

_Tip_: State will hold information about logged-in user. You can find `isAuthenticated` selector in `libs/core/src/lib/_states/auth.state.ts` that you can use.

### Challange 2:

One of our clients has hundreds or thousands of great ideas. Printing such a long list of ideas has a very bad impact on Angular performance. Could you please take a look on `IdeasListingComponent` (`apps/idea-board/src/app/ideas/ideas-list/_components/ideas-listing`) and check if we can easily optimise this component performance?

### Challange 3:

The dark mode is a must-have of all fancy apps those days, let's add it to our app.

_Tip_: All the logic is already there. You can start overriding styling in `apps/idea-board/src/styles/themes/_dark.scss`. Feel free to choose your own color scheme but if you do not have an idea for your own you can use those colors:

| Description                                               |  Color  |
| --------------------------------------------------------- | :-----: |
| App background                                            | #2f3337 |
| Component bg color (cards, login box, top navigation etc) | #3f4346 |
| Border color                                              | #828587 |
| Dark border color (eg idea card hover border color)       | #bbbbbb |
| Text color:                                               | #ffffff |
| Lighter text color (eg input placeholders):               | #bbbbbb |

### Challange 4:

Some say "it's a feature - not a bug", as good ideas should never be deleted, but it seems that you cannot remove idea from our Ideas board. Could you please take a look what's going on?

_Tip_: You will find the 'Delete' button if you go to edit mode for a particular idea. If you can't figure out where to start just open Dev tools and click that button.

### Challange 5:

We want to add some sorting/filtering functionality to the ideas board. Designs are not ready yet, but we can start working on some helper functions. Could you please finish all utilities function defined here `apps/idea-board/src/app/ideas/_utilities/ideas.utilities.ts` and add unit test for them?

_Tip_: Run `npm run affected:test --watch`.

### CODE REVIEW

File: front-end/README.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+### Challange 1:
+
+Oops, seems like you can access our idea list even if you are not logged in just by entering `http://localhost:4200/ideas`.
+This module should be accessed only by logged-in users. Could you please protect it somehow?
+
+_Tip_: State will hold information about logged-in user. You can find `isAuthenticated` selector in `libs/core/src/lib/_states/auth.state.ts` that you can use.
+
+### Challange 2:
+
+One of our clients have hundreds or thousends of great ideas. Printing so long list of idea have very bad inpact on Angular performance.
```
Comment (wojciech-szymanski) [inline front-end/README.md:54]:
```suggestion
One of our clients has hundreds or thousands of great ideas. Printing such a long list of ideas has a very bad impact on Angular performance.
```

File: front-end/README.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+Could you please take a look on `IdeasListingComponent` (`apps/idea-board/src/app/ideas/ideas-list/_components/ideas-listing`) and check
+if we can easly optimalise this component performance?
```
Comment (wojciech-szymanski) [inline front-end/README.md:56]:
```suggestion
if we can easily optimise this component performance?
```

File: front-end/apps/idea-board/src/app/ideas/_stores/ideas.data.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  {
+    id: 'idea-3',
+    name: 'Filtering',
+    description: 'Maybe we could add filtering by tags.',
+    tagIds: ['tag-3'],
+  },
+  {
+    id: 'idea-4',
+    name: 'Coffee break',
+    description: 'Coffee break increase developers performance',
```
Comment (wojciech-szymanski) [inline front-end/apps/idea-board/src/app/ideas/_stores/ideas.data.ts:26]:
```suggestion
    description: 'Coffee break increases developers performance',
```

File: front-end/apps/idea-board/src/app/ideas/_stores/ideas.data.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  {
+    id: 'idea-7',
+    name: 'Free fridays',
+    description: '... because why not?',
+    tagIds: ['tag-3'],
+  },
+  {
+    id: 'idea-8',
+    name: 'Mobile App',
+    description: `Let's build mobile app`,
```
Comment (wojciech-szymanski) [inline front-end/apps/idea-board/src/app/ideas/_stores/ideas.data.ts:51]:
```suggestion
    description: `Let's build a mobile app`,
```

File: front-end/apps/idea-board/src/styles/ui/_page-header.scss [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  &__title {
+    align-items: center;
+    display: flex;
+    flex-grow: 1;
+
+    > *:not(:first-child) {
+      margin-left: var(--spacing-small);
+    }
+  }
+  &__actions {
```
Comment (wojciech-szymanski) [inline front-end/apps/idea-board/src/styles/ui/_page-header.scss:23]:
```suggestion

  &__actions {
```

File: front-end/apps/idea-board/src/styles/ui/_page-header.scss [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    align-items: center;
+    display: flex;
+    justify-content: flex-end;
+
+    > *:not(:last-child) {
+      margin-right: var(--spacing-small);
+    }
+  }
+  &__intro {
```
Comment (wojciech-szymanski) [inline front-end/apps/idea-board/src/styles/ui/_page-header.scss:32]:
```suggestion

  &__intro {
```

File: front-end/apps/idea-board/src/styles/ui/_toggle.scss [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+      box-shadow: 0 0 0 1px rgba(0, 0, 0, 0.25);
+      width: 1rem;
+      height: 1rem;
+      position: absolute;
+      top: 0.25rem;
+      left: 0.25rem;
+      transition: left 0.25s;
+    }
+  }
+  &__checkbox {
```
Comment (wojciech-szymanski) [inline front-end/apps/idea-board/src/styles/ui/_toggle.scss:39]:
```suggestion

  &__checkbox {
```

**Review by wojciech-szymanski — CHANGES_REQUESTED** [review]
I have spotted some typos to fix and minor 💅 

Otherwise top job 👍

**Comment by gwidonw** [comment]
> 
> 
> I have spotted some typos to fix and minor 💅
> 
> Otherwise top job 👍

@wojciech-szymanski Many thanks. All suggestions were commited.

**Review by wojciech-szymanski — APPROVED** [review]

**Pull request merged**